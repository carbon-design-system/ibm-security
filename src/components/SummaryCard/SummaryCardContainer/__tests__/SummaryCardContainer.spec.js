/**
 * @file Summary card container tests.
 * @copyright IBM Security 2019
 */

import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React, { Fragment } from 'react';

import {
  SummaryCard,
  SummaryCardBatchAction,
  SummaryCardBatchActions,
  SummaryCardContainer,
  SummaryCardFooter,
  SummaryCardSelect,
} from '../../../..';

const summaryCards = [
  {
    id: '0',
  },
  {
    id: '1',
  },
];

describe('SummaryCardContainer', () => {
  test('should have no Axe or DAP violations', async () => {
    const main = document.createElement('main');
    render(
      <SummaryCardContainer
        render={({ getBatchActionProps, getSelectionProps, summaryCards }) => (
          <Fragment>
            <SummaryCardBatchActions {...getBatchActionProps()}>
              <SummaryCardBatchAction
                tabIndex={getBatchActionProps().shouldShowBatchActions ? 0 : -1}
              >
                test batch action
              </SummaryCardBatchAction>
            </SummaryCardBatchActions>
            {summaryCards.map(({ id }) => (
              <SummaryCard key={id}>
                <SummaryCardFooter>
                  <SummaryCardSelect
                    {...getSelectionProps({ id })}
                    id={`summary-select-${id}`}
                    labelText={`test select ${id}`}
                  />
                </SummaryCardFooter>
              </SummaryCard>
            ))}
          </Fragment>
        )}
        summaryCards={summaryCards}
      />,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );

    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations('SummaryCardContainer');
  });

  test('should have no Axe or DAP violations when cards are selected', async () => {
    const main = document.createElement('main');
    const { getByText } = render(
      <SummaryCardContainer
        render={({ getBatchActionProps, getSelectionProps, summaryCards }) => (
          <Fragment>
            <SummaryCardBatchActions {...getBatchActionProps()}>
              <SummaryCardBatchAction
                tabIndex={getBatchActionProps().shouldShowBatchActions ? 0 : -1}
              >
                test batch action
              </SummaryCardBatchAction>
            </SummaryCardBatchActions>
            {summaryCards.map(({ id }) => (
              <SummaryCard key={id}>
                <SummaryCardFooter>
                  <SummaryCardSelect
                    {...getSelectionProps({ id })}
                    id={`summary-select--${id}`}
                    labelText={`test select ${id}`}
                  />
                </SummaryCardFooter>
              </SummaryCard>
            ))}
          </Fragment>
        )}
        summaryCards={summaryCards}
      />,
      {
        // DAP requires a landmark '<main>' in the DOM:
        container: document.body.appendChild(main),
      }
    );

    // Currently there's no way to load the page
    // with summary cards pre-checked AND batch action toolbar updated.
    // Therefore, we have to select both cards first:
    userEvent.click(getByText(/test select 0/i));
    userEvent.click(getByText(/test select 1/i));

    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations(
      'SummaryCardContainer with selected cards'
    );
  });

  test('should cycle elements in tab order', () => {
    const { getByLabelText } = render(
      <SummaryCardContainer
        render={({ getBatchActionProps, getSelectionProps, summaryCards }) => (
          <Fragment>
            <SummaryCardBatchActions {...getBatchActionProps()}>
              <SummaryCardBatchAction
                tabIndex={getBatchActionProps().shouldShowBatchActions ? 0 : -1}
              >
                test batch action
              </SummaryCardBatchAction>
            </SummaryCardBatchActions>
            {summaryCards.map(({ id }) => (
              <SummaryCard key={id}>
                <SummaryCardFooter>
                  <SummaryCardSelect
                    {...getSelectionProps({ id })}
                    id={`summary-select--${id}`}
                    labelText={`test select ${id}`}
                  />
                </SummaryCardFooter>
              </SummaryCard>
            ))}
          </Fragment>
        )}
        summaryCards={summaryCards}
      />
    );

    userEvent.tab();

    // The first summary card select:
    expect(getByLabelText(/test select 0/i)).toHaveFocus();

    userEvent.tab();

    // Loop complete.
    // The first summary card select:
    expect(getByLabelText(/test select 1/i)).toHaveFocus();
  });

  test('should cycle elements in tab order when the batch actions toolbar is activated', () => {
    const { getByLabelText, getByText } = render(
      <SummaryCardContainer
        render={({ getBatchActionProps, getSelectionProps, summaryCards }) => (
          <Fragment>
            <SummaryCardBatchActions {...getBatchActionProps()}>
              <SummaryCardBatchAction
                tabIndex={getBatchActionProps().shouldShowBatchActions ? 0 : -1}
              >
                test batch action
              </SummaryCardBatchAction>
            </SummaryCardBatchActions>
            {summaryCards.map(({ id }) => (
              <SummaryCard key={id}>
                <SummaryCardFooter>
                  <SummaryCardSelect
                    {...getSelectionProps({ id })}
                    id={`summary-select--${id}`}
                    labelText={`test select ${id}`}
                  />
                </SummaryCardFooter>
              </SummaryCard>
            ))}
          </Fragment>
        )}
        summaryCards={summaryCards}
      />
    );

    // Currently there's no way to load the page
    // with summary cards pre-checked AND batch action toolbar updated.
    // Therefore, we have to select both cards first:
    userEvent.click(getByText(/test select 0/i));
    userEvent.click(getByText(/test select 1/i));

    userEvent.tab();

    // The batch action button:
    expect(getByText(/test batch action/i)).toHaveFocus();

    userEvent.tab();

    // The batch actions cancel button:
    expect(getByText(/Cancel/i)).toHaveFocus();

    userEvent.tab();

    // The first summary card select:
    expect(getByLabelText(/test select 0/i)).toHaveFocus();

    userEvent.tab();

    // The first summary card select:
    expect(getByLabelText(/test select 1/i)).toHaveFocus();

    userEvent.tab();

    // Loop complete.
    // The batch action button:
    expect(getByText(/test batch action/i)).toHaveFocus();
  });

  test('should update batch actions when a summary card is selected', () => {
    const { getByLabelText, queryByText } = render(
      <SummaryCardContainer
        render={({ getBatchActionProps, getSelectionProps, summaryCards }) => (
          <Fragment>
            <SummaryCardBatchActions {...getBatchActionProps()}>
              <SummaryCardBatchAction
                tabIndex={getBatchActionProps().shouldShowBatchActions ? 0 : -1}
              >
                test batch action
              </SummaryCardBatchAction>
            </SummaryCardBatchActions>
            {summaryCards.map(({ id }) => (
              <SummaryCard key={id}>
                <SummaryCardFooter>
                  <SummaryCardSelect
                    {...getSelectionProps({ id })}
                    id={`summary-select--${id}`}
                    labelText={`test select ${id}`}
                  />
                </SummaryCardFooter>
              </SummaryCard>
            ))}
          </Fragment>
        )}
        summaryCards={summaryCards}
      />
    );

    // We can't check `not.toBeVisible` because the batch actions
    // are absolutely positioned & not hidden/removed from the DOM.
    // So instead, we query the default batch actions text with 0 selected:
    expect(queryByText(/0 item selected/i)).toBeInTheDocument();

    userEvent.click(getByLabelText(/test select 0/i));

    // Expect first summary card to be selected:
    expect(getByLabelText(/test select 0/i)).toBeChecked();

    // Expect the batch actions text to be updated:
    expect(queryByText(/1 item selected/i)).toBeInTheDocument();
  });

  test('should clear batch selections when the cancel button is clicked', () => {
    const { getByLabelText, getByText, queryByText } = render(
      <SummaryCardContainer
        render={({ getBatchActionProps, getSelectionProps, summaryCards }) => (
          <Fragment>
            <SummaryCardBatchActions {...getBatchActionProps()}>
              <SummaryCardBatchAction
                tabIndex={getBatchActionProps().shouldShowBatchActions ? 0 : -1}
              >
                test batch action
              </SummaryCardBatchAction>
            </SummaryCardBatchActions>
            {summaryCards.map(({ id }) => (
              <SummaryCard key={id}>
                <SummaryCardFooter>
                  <SummaryCardSelect
                    {...getSelectionProps({ id })}
                    id={`summary-select--${id}`}
                    labelText={`test select ${id}`}
                  />
                </SummaryCardFooter>
              </SummaryCard>
            ))}
          </Fragment>
        )}
        summaryCards={summaryCards}
      />
    );

    // Currently there's no way to load the page
    // with summary cards pre-checked AND batch action toolbar updated.
    // Therefore, we have to select both cards first:
    userEvent.click(getByText(/test select 0/i));
    userEvent.click(getByText(/test select 1/i));

    // Expect both summary cards to be selected at first:
    expect(queryByText(/2 items selected/i)).toBeInTheDocument();

    // Clear batch actions selections:
    userEvent.click(getByText(/Cancel/i));

    // Expect the batch actions text to be updated:
    expect(queryByText(/0 item selected/i)).toBeInTheDocument();

    // Expect both summary cards to not be selected:
    expect(getByLabelText(/test select 0/i)).not.toBeChecked();
    expect(getByLabelText(/test select 1/i)).not.toBeChecked();
  });
});
