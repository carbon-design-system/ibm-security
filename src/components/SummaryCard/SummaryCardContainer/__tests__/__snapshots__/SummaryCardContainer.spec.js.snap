// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryCardContainer Rendering renders 1`] = `
<div
  class="security--summary-card__container"
>
  <div
    class="bx--batch-actions security--summary-card__batch-actions"
    data-testid="batchActions"
  >
    <div
      class="bx--action-list"
    >
      <button
        class="bx--btn bx--btn--primary"
        tabindex="0"
        type="button"
      >
        SummaryCardBatchAction
        <svg
          aria-hidden="true"
          class="bx--btn__icon"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          style="will-change: transform;"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 1C4.2 1 1 4.2 1 8s3.2 7 7 7 7-3.2 7-7-3.2-7-7-7zm4 8H9v3H7V9H4V7h3V4h2v3h3v2z"
          />
        </svg>
      </button>
      <button
        class="bx--batch-summary__cancel bx--btn bx--btn--primary"
        tabindex="0"
        type="button"
      >
        Cancel
      </button>
    </div>
    <div
      class="bx--batch-summary"
    >
      <p
        class="bx--batch-summary__para"
      >
        <span>
          0 item selected
        </span>
      </p>
    </div>
  </div>
  <section
    class="security--summary-card"
  >
    <div
      class="security--summary-card__footer"
    >
      <div
        class="security--summary-card__select"
      >
        <div
          class="bx--form-item bx--checkbox-wrapper"
        >
          <input
            class="bx--checkbox"
            data-testid="select__summary-card__0"
            id="security--summary-card__select__summary-card__0"
            type="checkbox"
          />
          <label
            class="bx--checkbox-label security--summary-card__select__checkbox"
            for="security--summary-card__select__summary-card__0"
          >
            <span
              class="bx--checkbox-label-text"
            >
              Select
            </span>
          </label>
        </div>
      </div>
    </div>
  </section>
  <section
    class="security--summary-card"
  >
    <div
      class="security--summary-card__footer"
    >
      <div
        class="security--summary-card__select"
      >
        <div
          class="bx--form-item bx--checkbox-wrapper"
        >
          <input
            class="bx--checkbox"
            data-testid="select__summary-card__1"
            id="security--summary-card__select__summary-card__1"
            type="checkbox"
          />
          <label
            class="bx--checkbox-label security--summary-card__select__checkbox"
            for="security--summary-card__select__summary-card__1"
          >
            <span
              class="bx--checkbox-label-text"
            >
              Select
            </span>
          </label>
        </div>
      </div>
    </div>
  </section>
  <section
    class="security--summary-card"
  >
    <div
      class="security--summary-card__footer"
    >
      <div
        class="security--summary-card__select"
      >
        <div
          class="bx--form-item bx--checkbox-wrapper"
        >
          <input
            class="bx--checkbox"
            data-testid="select__summary-card__2"
            id="security--summary-card__select__summary-card__2"
            type="checkbox"
          />
          <label
            class="bx--checkbox-label security--summary-card__select__checkbox"
            for="security--summary-card__select__summary-card__2"
          >
            <span
              class="bx--checkbox-label-text"
            >
              Select
            </span>
          </label>
        </div>
      </div>
    </div>
  </section>
  <section
    class="security--summary-card"
  >
    <div
      class="security--summary-card__footer"
    >
      <div
        class="security--summary-card__select"
      >
        <div
          class="bx--form-item bx--checkbox-wrapper"
        >
          <input
            class="bx--checkbox"
            data-testid="select__summary-card__3"
            id="security--summary-card__select__summary-card__3"
            type="checkbox"
          />
          <label
            class="bx--checkbox-label security--summary-card__select__checkbox"
            for="security--summary-card__select__summary-card__3"
          >
            <span
              class="bx--checkbox-label-text"
            >
              Select
            </span>
          </label>
        </div>
      </div>
    </div>
  </section>
  <section
    class="security--summary-card"
  >
    <div
      class="security--summary-card__footer"
    >
      <div
        class="security--summary-card__select"
      >
        <div
          class="bx--form-item bx--checkbox-wrapper"
        >
          <input
            class="bx--checkbox"
            data-testid="select__summary-card__4"
            id="security--summary-card__select__summary-card__4"
            type="checkbox"
          />
          <label
            class="bx--checkbox-label security--summary-card__select__checkbox"
            for="security--summary-card__select__summary-card__4"
          >
            <span
              class="bx--checkbox-label-text"
            >
              Select
            </span>
          </label>
        </div>
      </div>
    </div>
  </section>
</div>
`;
